spark/src/main/scala/geotrellis/spark/pyramid/Pyramid.scala

```diff
@@ -42,6 +65,69 @@ object Pyramid extends LazyLogging {
     implicit def methodToOptions(m: ResampleMethod): Options = Options(resampleMethod = m)
   }
 
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag: SpatialComponent,
+    V <: CellGrid: ? => TilePrototypeMethods[V]: ? => TileMergeMethods[V]: AvroRecordCodec: ClassTag,
+    M: JsonFormat: Component[?, Bounds[K]]: Component[?, LayoutDefinition]
+  ](layerName: String, maxZoom: Int, minZoom: Int, layerReader: LayerReader[LayerId]): Pyramid[K, V, M] = {
+    val seq = for (z <- maxZoom to minZoom by -1) yield {
+      (z, layerReader.read[K, V, M](LayerId(layerName, z)))
+    }
+    Pyramid[K, V, M](seq.toMap)
+  }
+
+  def write[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag: SpatialComponent,
+    V <: CellGrid: ? => TilePrototypeMethods[V]: ? => TileMergeMethods[V]: AvroRecordCodec: ClassTag,
+    M: JsonFormat: Component[?, Bounds[K]]: Component[?, LayoutDefinition]
+  ](pyramid: Pyramid[K, V, M], layerName: String, writer: LayerWriter[LayerId], keyIndexMethod: KeyIndexMethod[K]) = {
+    for (z <- pyramid.maxZoom to pyramid.minZoom by -1) {
+      writer.write[K, V, M](LayerId(layerName, z), pyramid(z), keyIndexMethod)
+    }
+  }
+
+  def apply[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: ? => TilePrototypeMethods[V]: ? => TileMergeMethods[V],
+    M: Component[?, LayoutDefinition]: Component[?, Bounds[K]]
+  ](rdd: RDD[(K, V)] with Metadata[M],
+    startZoom: Int,
+    endZoom: Int,
+    options: Options
+  ): Pyramid[K, V, M] = {
+    Pyramid(levelStream(rdd, new LocalLayoutScheme, startZoom, endZoom, options).toMap)
+  }
+
+  def apply[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: ? => TilePrototypeMethods[V]: ? => TileMergeMethods[V],
+    M: Component[?, LayoutDefinition]: Component[?, Bounds[K]]
+  ](rdd: RDD[(K, V)] with Metadata[M],
+    startZoom: Int,
+    endZoom: Int
+  ): Pyramid[K, V, M] =
+    apply(rdd, startZoom, endZoom, Options.DEFAULT)
+
+  def apply[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: ? => TilePrototypeMethods[V]: ? => TileMergeMethods[V],
+    M: Component[?, LayoutDefinition]: Component[?, Bounds[K]]
+  ](rdd: RDD[(K, V)] with Metadata[M],
+    options: Options
+  ): Pyramid[K, V, M] = {
+    val gridBounds = rdd.metadata.getComponent[Bounds[K]].asInstanceOf[KeyBounds[K]].toGridBounds
+    val maxDim = math.max(gridBounds.width, gridBounds.height).toDouble
+    val levels = math.ceil(math.log(maxDim)/math.log(2)).toInt
+
+    apply(rdd, 0, levels, options)
+  }
+
+  def apply[
```

Comment:
I think these overloads can be actually collapsed into a single method with defaults.